### PR TITLE
Explain the current state of our SEA work

### PR DESCRIPTION
**Pull request by jviotti**
This document is a bit like blog post, in that in documents the state at
this point in time and will likely be out of date at some point. For
this reason, I put this document inside a `blog/` directory, prefixed
with its date.

Signed-off-by: Juan Cruz Viotti <jv@jviotti.com>

### CODE REVIEW

File: notes/2022-08-05-an-overview-of-the-current-state.md [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+project.
+
+In the area of *Virtual File Systems*, we are currently basing our
+proof-of-concept on the [ASAR](https://github.com/electron/asar) archive format
+designed and battle-tested by Electron. The ASAR format is extensible by
+design, allowing us to add arbitrary new metadata and functionality to our
+custom implementation.
+
+In the area of the *Bootstrapper*, we are maintaining a custom Node.js patch
+that makes use of Postject relfection APIs, provides a basic ASAR read-only
```
Comment (robertgzr) [inline notes/2022-08-05-an-overview-of-the-current-state.md:233]:
```suggestion
that makes use of Postject reflection APIs, provides a basic ASAR read-only
```

File: notes/2022-08-05-an-overview-of-the-current-state.md [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+Future work
+-----------
+
+SEAs open interesting possibilities for innovation. For example, we can explore
+creating SEAs that make use of v8 snapshots to speed up application startup
+time, or ways in which we can trim down Node.js to optimize for
+space-efficiency in the context of SEAs.
+
+Idea are very welcomed!
```
Comment (robertgzr) [inline notes/2022-08-05-an-overview-of-the-current-state.md:249]:
```suggestion
Ideas are very welcome!
```

Comment (jviotti) [reply notes/2022-08-05-an-overview-of-the-current-state.md]: Done!

File: notes/2022-08-05-an-overview-of-the-current-state.md (hunk 1 above)
Comment (jviotti) [reply notes/2022-08-05-an-overview-of-the-current-state.md]: Done!

File: notes/2022-08-05-an-overview-of-the-current-state.md [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+The initial set of requirements for this ingredient are:
+
+| # | Requirement                                                                        |
+|---|------------------------------------------------------------------------------------|
+| 1 | Inject data within the boundaries of the binary format, not outside of it          |
+| 2 | Support every binary format adopted by the platforms supported by Node.js          |
+| 3 | Support the injection of any arbitrary data, irrelevant of its format and contents |
+| 4 | Provide a complementary cross-platform native API for runtime reflection           |
+| 5 | Should be a pure JavaScript solution or not require native add-ons                 |
```
Comment (saswatds) [inline notes/2022-08-05-an-overview-of-the-current-state.md:145]: Is the uses of only javascript for the solution a hard requirement?

**Comment by guest271314** [comment]
> Node.js executable may seem daunting at first, but its not!

"its" should be "it's".

> First, we start with the assets that make a typical Node.js application. These
typically consist of JavaScript files from the application code

"typically" is redundant in 2d sentence, where "typical" is used in 1st sentence.

> Most interesting Node.js programs are composed by more than one file.

Is this opinion or fact? There should not be a difference from the user viewpoint between creating a SEA with 1 or 100 input files.

> - **Require maintaining custom Node.js patches.** These patches typically touch
  on the initialization logic of Node.js and are different across Node.js
  versions. These patches result in high maintenance burden and make it
  difficult for SEA implementations to support new Node.js versions as soon as
  they are released.

This is what I was alluding to in my posts in https://github.com/nodejs/single-executable/issues/1. That is, the process should
be as simple as a user visisting the Node.js downloads Website, uploading their existsing Node.js-specific JavaScript
code and the site producing out an executable with tip-of-tree (nightly) node executable for download. That way the SEA
is always up to date, and/or is capable of being re-compiled faily/nightly. Perhaps that is a point
that needs to be coordinated with the team/WG that published node nightly.

> **No support for non-JavaScript assets.** Some SEA implementations
  concatenate JavaScript assets and inject the resulting "single" file into the
  executable. While this approach works for simple programs, many Node.js
  applications require application resources that are not typically
  concatenated as part of the JavaScript code, such as Node.js native add-ons,
  text files, executable scripts and more. These assets are often dynamically
  consumed using `fs` and `child_process`.

I'll point out here that QuickJS has its own compiler, `qjsc` https://bellard.org/quickjs/quickjs.html#qjsc-compiler.

**Comment by Trott** [comment]
> > Node.js executable may seem daunting at first, but its not!
> 
> "its" should be "it's".

Please use [GitHub's review comment feature](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request) for comments like this. Even better if you use their suggestion feature so someone can simply commit your suggestion within the interface.

**Comment by guest271314** [comment]
I thought I did. Then couldn't find the same commit again until someone a collaborator/member replied to my suggestion to remove the bit about "interesting" https://github.com/nodejs/single-executable/commit/2f5c6ac974a9c970178dd4e9d52f8af432f91dd3. Done.

File: notes/2022-08-05-an-overview-of-the-current-state.md [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+We believe that joining forces will not only result in a superior solution in
+the context of Node.js, but also introduce foundational blocks to solve related
+problems in the open-source world at large.
+
+How do SEAs work?
+-----------------
+
+The problem of combining application resources with a Node.js executable may
+seem daunting at first, but its not! The process typically looks something like
```
Comment (Trott) [inline notes/2022-08-05-an-overview-of-the-current-state.md:41]:
```
seem daunting at first, but it's not! The process typically looks something like
```

File: notes/2022-08-05-an-overview-of-the-current-state.md [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+The initial set of requirements for this ingredient are:
+
+| # | Requirement                                                                        |
+|---|------------------------------------------------------------------------------------|
+| 1 | Inject data within the boundaries of the binary format, not outside of it          |
+| 2 | Support every binary format adopted by the platforms supported by Node.js          |
+| 3 | Support the injection of any arbitrary data, irrelevant of its format and contents |
+| 4 | Provide a complementary cross-platform native API for runtime reflection           |
+| 5 | Should be a pure JavaScript solution or not require native add-ons                 |
+| 6 | Allow injection on any supported binary format in any platform                     |
```
[the remark this answers is not in the document]
Comment (jviotti) [reply notes/2022-08-05-an-overview-of-the-current-state.md:146]: Point 2 talks about the capability of injecting data in each platform. For example, using Windows to inject the data into an EXE, using macOS to inject the data into a Mach-O, etc.

Point 6 is about being able to do any injection on any operating system. For example, being able to use macOS to produce a Windows SEA, etc.

We support this at the moment on Postject thanks to LIEF, and its pretty handy for i.e. continuous integration purposes, as you can generate all binaries for your app through a single Linux runner, etc.

I'll try to clarify this better now.

File: notes/2022-08-05-an-overview-of-the-current-state.md (hunk 3 above)
Comment (jviotti) [reply notes/2022-08-05-an-overview-of-the-current-state.md:145]: It was framed as a hard requirement by @jesec based on his experience managing PKG. A lot of people are likely to face troubles setting things up locally (i.e. mainly on Windows), resulting in unnecessary maintenance overhead.

Maybe "pure JavaScript" is not the right term... Something like WebAssembly would do as well. I'll clarify

**Comment by jviotti** [comment]
@guest271314 Thanks for the suggestions!

> Is this opinion or fact? There should not be a difference from the user viewpoint between creating a SEA with 1 or 100 input files.

Very true. What I wanted to exemplify is that a notion of a VFS is often needed for apps consisting of more than one file (which might not even be `require()`-able files or are dynamically loaded). A VFS is not very useful for an app built as a single JS file.

> This is what I was alluding to in my posts in https://github.com/nodejs/single-executable/issues/1. That is, the process should be as simple as a user visisting the Node.js downloads Website, uploading their existsing Node.js-specific JavaScript code and the site producing out an executable with tip-of-tree (nightly) node executable for download. That way the SEA is always up to date, and/or is capable of being re-compiled daily/nightly. Perhaps that is a point that needs to be coordinated with the team/WG that published node nightly.

Exactly. Projects like PKG involve significant maintenance overhead to produce Node.js binaries that can injected into. If we can provide that capability with upstream Node.js builds, then this entire complexity goes away! I tried to clarify this better now on the "Problems with current SEAs" section.

> I'll point out here that QuickJS has its own compiler, qjsc https://bellard.org/quickjs/quickjs.html#qjsc-compiler.

Thanks for sharing! We should definitely document this

File: notes/2022-08-05-an-overview-of-the-current-state.md (hunk 4 above)
Comment (jviotti) [reply notes/2022-08-05-an-overview-of-the-current-state.md]: Fixed!

**Review by RaisinTen — COMMENTED** [review]
> This document is a bit like blog post

Should we place this in a directory called `blog` then?

File: notes/2022-08-05-an-overview-of-the-current-state.md [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+The initial set of requirements for this ingredient are:
+
+| # | Requirement                                                             |
+|---|-------------------------------------------------------------------------|
+| 1 | Support random access reads for performance reasons                     |
+| 2 | Support the concept of symbolic links                                   |
+| 3 | Support preserving file permissions, at least the executable bit        |
+| 4 | Support general purpose data compression for space optimization reasons |
+| 5 | Preserve file-hierarchy information                                     |
+| 6 | Increase locality of related files for performance reasons              |
```
[the remark this answers is not in the document]
Comment (jviotti) [reply notes/2022-08-05-an-overview-of-the-current-state.md:185]: Makes sense! This document is supposed to be a high-level explanation of the problem space. I hope to collect proper requirements for each component separately, and we could go deeper in the details you are mentioning here then!

File: notes/2022-08-05-an-overview-of-the-current-state.md [hunk 7]
```diff
@@ -0,0 +1,283 @@
+An Overview of the Current State
+================================
+
+Authors: [@jviotti](https://github.com/jviotti)
```
[the remark this answers is not in the document]
Comment (jviotti) [reply notes/2022-08-05-an-overview-of-the-current-state.md:4]: Yep, let's do it! Added the key authors + reviewers in alphabetical order.

**Comment by jviotti** [comment]
> Should we place this in a directory called blog then?

Works for me! Done!

**Comment by guest271314** [comment]
Doesn't Native Client do something like what is being worked on here? 

I tried to reverse-engineer a .nexe file; contains the entire program in one file.

File: notes/2022-08-05-an-overview-of-the-current-state.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (RaisinTen) [reply notes/2022-08-05-an-overview-of-the-current-state.md]: :+1:
I believe this change got removed in your last force push?

[the remark this answers is not in the document]
Comment (jviotti) [reply notes/2022-08-05-an-overview-of-the-current-state.md]: My bad. Restored!

**Comment by jviotti** [comment]
I don't know a lot about Native Client. Can you open a GitHub Discussion on this repo about it so we get more eyes on it? It'd be great if you can explain in more detail what it does (for people who are not super familiar) and even show some of the results you've seen from reverse engineering that `.nexe` file.
 
According to https://developer.chrome.com/docs/native-client/welcome-to-native-client/, it is deprecated technology for anything other than ChromeOS, but there might definitely be some lessons we can learn from it.

**Review by RaisinTen — APPROVED** [review]
LGTM in its current state. Would anyone else from @nodejs/single-executable like to drop your LGTMs / request changes on this one?

**Comment by jviotti** [comment]
Awesome. I'll merge now so we can get the ball rolling, but happy to address other comments on further PRs.

**Pull request merged**